RFD 0119 - Integrate with AWS using OIDC (#22556)
gravitational/teleport: rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md

---
authors: Marco Dinis (marco.dinis@goteleport.com)
state: draft
---

# RFD 999 - OIDC IdP for AWS RDS DBs discovery

## Required Approvals

* Engineering: @ryanclark && @r0mant
* Product: @xinding33 || @klizhentas
* Security: @reedloden

## What

Discover AWS RDS Databases without manual configuration.

#### Goals

> Easily discover AWS RDS Databases of an account.

Author: I was targeting something more specific to avoid the "where are we going to use this"

Let me rephrase the goal and RFD title to "Integrate with AWS using OIDC"

#### Non-goals

Having a real OIDC IdP is out of scope.
That initiative can be tracked [here](https://github.com/gravitational/teleport/issues/20967).

Automatically onboard AWS RDS Databases into Teleport.

Discovering other AWS resources is out of scope.

## Terminology

* OIDC: OpenID Connect is a protocol that works on top of OAuth2 and provides identity information.
* IdP: Identity Provider is a system that manages identities.
* JWT: JSON Web Token is a json encoded token that contains a set of claims signed by an entity.
* JWKS: JSON Web Key Set is a set of public keys that IdP consumers must use to validate a signed JWT.

## Why

Our current discovery process for AWS resources requires multiple steps before the user gets to the "aha moment".
Some of those steps are a little complicated and require the user to know some Teleport specific internals.

> We can reduce the number of steps and remove most of the Teleport specific configuration by creating an OIDC IdP in Teleport.

Author: :+1:
Mentioning cloud here seems out of place :thinking: 

I'll add it to the goals instead, if that's ok

## How

AWS allows the [set up](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) of an IdP using OIDC providers.
Each provider has a set of roles, which limit their access.

Turning Teleport into an OIDC IdP, and asking the user to create a new OIDC IdP that consumes the Teleport OIDC endpoints, allows us to easily call AWS API endpoints to, for instance, list RDS instances and their details.

When configuring the provider, we'll need an AWS Role which Teleport uses to issue API Calls.

> To store the configuration above, we'll create a new resource Kind: `Integration`.

Reviewer: Can users manage these integrations dynamically with tctl or API? I assume they can, let's include some examples.

Author: I was thinking of only supporting via API, but I guess we should add `tctl` as well (it's always handy and not very expensive)

It'll leverage the `subkind` prop to distinguish future integrations.

```yaml
kind: integration

> subkind: aws-oidc

Reviewer: Nit: Do we need `-oidc` suffix?

Author: I'm trying to leave space/pattern to have other types of AWS integrations (using secrets or using an ec2 instance).

Should I remove the `-oidc` anyway?

version: v1
metadata:
	name: some-name
spec:

> 	aws_role: <aws role>

Reviewer: Nit:
```suggestion
	aws_role_arn: <aws role>
```

#### System's Point of View
The first interaction between AWS and Teleport happens when the user clicks on `Get Thumbprint` (step 4 of the User's Point of View flow).
This will trigger a request started by AWS to Teleport hitting the OpenID Configuration endpoint and subsequently the JSON Web Key Set endpoint.

After this step, when the user tries to list RDS DBs, Teleport generates a token and issues the API call.

To generate a token, Teleport creates a JWT with the claims described above and the configured role, and signs it with the private key (the public is provided in the public JWKS endpoint).

AWS receives the request and validates the token against the public key provided by OIDC IdP (ie, Teleport) at the `jwks_uri`.
If authenticated and authorized for the api call, AWS returns the response to the API call.

### Implementation Details

#### Signing Key
One of the requirements to be an OIDC provider is to provide the public key in a known HTTP endpoint and sign a JSON object (with claims) with the private key.

We'll create a new RSA 2048 Key type and add it to the CertAuthority.

> This will be very similar to the JWTSigner key type used for App access.

Reviewer: So are we going to introduce another CA type? Like, `OIDCCA`? Mention it explicitly in the document.

Reviewer 2: I'd be keen for us to introduce a CA just for issuing OIDC ID Tokens

A single signing key will be used for this flow even if multiple AWS integrations are created (eg, multiple regions).

#### HTTP Endpoints
The following endpoints will be used during OIDC IdP set up:

##### OpenID Configuration
According to the [spec](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig) the Identity Provider must provide an endpoint at `<providerURL>/.well-known/openid-configuration` that returns the provider's configuration.

So, a new endpoint at `<teleportProviderURL>/.well-known/openid-configuration` will be created that returns the following JSON:
```
200 OK

{
  "issuer": "<teleportProviderURL>", 
  "jwks_uri": "<teleportProviderURL>/.well-known/jwks-oidc",
  "claims": ["iss", "sub", "aud", "jti", "iat", "exp", "nbf"],
  "id_token_signing_alg_values_supported": ["RS256"],
  "response_types_supported": ["id_token"],
  "scopes_supported": ["openid"],
  "subject_types_supported": ["public", "pairwise"]
}
```

> - Issuer: the provider's URL

Reviewer: This is basically proxy's public address, correct? Let's just call it that explicitly so it's clear.

- JWKS URI: the endpoint where the provider returns the public keys.
- Claims Supported: a list of supported claims that the OpenID Provider (ie, Teleport) MAY be able to supply values for when issuing a token:
  - `iss`: the issuer identity, on our case it will be the same as the `issuer` key.

>   - `sub`: identifies the subject of the token, on our case it will contain the user's name.

Reviewer: ```suggestion
  - `sub`: identifies the subject of the token, in our case it will contain the Teleport username.
```

>   - `sub`: identifies the subject of the token, on our case it will contain the user's name.

Author: Sure, seems fair :+1:

  - `aud`: the audience for the token, on our case it will be the same as the `audience` defined when configuring the OIDC IdP in AWS (step 3.3 for the User's Point of View).

>   - `jti`: is the token id.

Reviewer: UUIDv4? Mention that.

```
200 OK

{
  "keys": [
    {
      "kty": "RSA",
      "alg": "RS256",
      "n": "<public key value>",
      "e": "AQAB"
    }
  ]
}
```

This must respect the RFC7517 (eg of a key can be found at https://www.rfc-editor.org/rfc/rfc7517#appendix-A.1).

Obtaining this key should be possible using
```golang
GetCertAuthority(
	ctx,
	types.CertAuthID{
		Type:       types.OIDCSigner, // new key type
		DomainName: clusterName,
	},
	false /*loadKeys*/,
)
```

#### AWS Role for Teleport OIDC IdP

The user will create or associate an AWS Role to the Teleport OIDC IdP.
This role must have access to list RDS Databases and have a policy that trusts Teleport as an Identity Provider.

At least a policy allowing the following must be part of the role:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "rds:DescribeDBInstances",
            "Resource": "*"
        }
    ]
}
```

The AWS Role must trust Teleport as an IdP.
To do so, the user must add the following Trusted relationship:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {

>                 "Federated": "arn:aws:iam::<account id>:oidc-provider/<teleportProviderURL>"

Reviewer: Do you embed the actual proxy address here and below? Can you use specific values for better demonstration? E.g. say we have a proxy at `https://proxy.example.com:3080`.

            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "<teleportProviderURL>:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```

### Security

#### Rotation
JWT Signing keys can be rotated using `tctl auth rotate --type=oidc`.

JWKS endpoint returns a list of public keys, so the old and the new key are provided for a seamless migration.

#### MITM between AWS and Teleport

The only configuration AWS has about the provider is its URL.

> If that DNS is controlled by some evil part, AWS might accept requests from that evil part.

Reviewer: ```suggestion
If DNS is controlled by an attacker, AWS might accept requests from them.
```

However, assuming the DNS is controlled by an evil part, then the users that are logging in to Teleport are also providing their credentials to that part.

We don't think this is a scenario we should focus on this RFD.

#### AWS Calls from unauthorized Teleport Users
We'll be able to call `DescribeDBInstances` AWS endpoint, however we must ensure that this is protected by RBAC.

> In order to do so, we'll re-use the `db` resource with the `list` verb to allow listing RDS Databases.

Author: Picture the state where we have an Integration with AWS.
We'll have an HTTP endpoint in Proxy that calls the AWS API Describe DB Instance.
We'll have a new resource Integration, and subsequent rules (create, delete, list, update), but that only works for CRUD operations over the resource itself.

Should any user be able to list the AWS RDS DBs? Or do we want some kind of check before calling the AWS Endpoint?
I'm overloading the `db.list` rule with something else not entirely relevant.
However, I think we should add some check before calling the AWS Endpoint.

> In order to do so, we'll re-use the `db` resource with the `list` verb to allow listing RDS Databases.

Reviewer 3: > We'll have a new verb: `use` If the user has `integration.use` then they can issue Integration API Calls

Apologies for a naive, after-the-merge question, but what is stopping us from using one of the existing verbs like "read" for this? Does it have to do with the part where you mention rules working only for CRUD operations over the resource itself?

We have managed to avoid adding a verb like "use" for resources where one could theoretically make use of it as well ("using" a db server or a node) so I'm wondering how this situation is different.

If it's related to the `POST .../integration/:id/action/aws-oidc-list-databases` endpoint mentioned in the RFD, then we could even argue that describing access to call an integration action could be made through checking the "create" verb for a new resource called "integration action". OTOH I've seen someone mention that it's hard to implement those checks for "virtual" resources that don't actually exist on the backend.

> In order to do so, we'll re-use the `db` resource with the `list` verb to allow listing RDS Databases.

Author: There's some discussion here, which provide some context
https://github.com/gravitational/teleport/pull/24256#discussion_r1163043822

> We have managed to avoid adding a verb like "use" for resources where one could theoretically make use of it as well ("using" a db server or a node) so I'm wondering how this situation is different.

They are protected by `<resource>_labels`, so we could also use that here.
We discussed that approach but ended deciding on using the `VerbUse`

As for creating a virtual resource `integration action`, I don't think it was considered.
I'm not aware of any issues of implementing resources without a backend. But I didn't try it.

I think it could be an alternative, but don't actually see a big disadvantage in using the `VerbUse`
